x-pack/apm-server/aggregation/servicesummarymetrics/aggregator.go

```diff
@@ -0,0 +1,264 @@
+	}
+	return nil
+}
+
+// Aggregator aggregates all events, periodically publishing service summary metrics.
+type Aggregator struct {
+	*baseaggregator.Aggregator
+
+	config AggregatorConfig
+
+	mu sync.RWMutex
+	// These two metricsBuffer are set to the same size and act as buffers
+	// for caching and then publishing the metrics as batches.
+	active, inactive map[time.Duration]*metricsBuffer
+}
+
+// NewAggregator returns a new Aggregator with the given config.
+func NewAggregator(config AggregatorConfig) (*Aggregator, error) {
+	if err := config.Validate(); err != nil {
+		return nil, errors.Wrap(err, "invalid aggregator config")
+	}
+	if config.Logger == nil {
+		config.Logger = logp.NewLogger(logs.ServiceSummaryMetrics)
+	}
+	aggregator := Aggregator{
+		config:   config,
+		active:   make(map[time.Duration]*metricsBuffer),
+		inactive: make(map[time.Duration]*metricsBuffer),
+	}
+	base, err := baseaggregator.New(baseaggregator.AggregatorConfig{
+		PublishFunc:     aggregator.publish, // inject local publish
+		Logger:          config.Logger,
+		Interval:        config.Interval,
+		RollUpIntervals: config.RollUpIntervals,
+	})
+	if err != nil {
+		return nil, err
+	}
+	aggregator.Aggregator = base
+	for _, interval := range aggregator.Intervals {
+		aggregator.active[interval] = newMetricsBuffer(config.MaxGroups)
+		aggregator.inactive[interval] = newMetricsBuffer(config.MaxGroups)
+	}
+	return &aggregator, nil
+}
+
+func (a *Aggregator) publish(ctx context.Context, period time.Duration) error {
+	// We hold a.mu only long enough to swap the serviceSummaryMetrics. This will
+	// be blocked by serviceSummaryMetrics updates, which is OK, as we prefer not
+	// to block serviceSummaryMetrics updaters. After the lock is released nothing
+	// will be accessing a.inactive.
+	a.mu.Lock()
+
+	// EXPLAIN: We swap active <-> inactive, so that we're only working on the
+	// inactive property while publish is running. `a.active` is the buffer that
+	// receives/stores/updates the metricsets, once swapped, we're working on the
+	// `a.inactive` which we're going to process and publish.
+	current := a.active[period]
+	a.active[period], a.inactive[period] = a.inactive[period], current
+	a.mu.Unlock()
+
+	size := len(current.m)
+	if size == 0 {
+		a.config.Logger.Debugf("no service summary metrics to publish")
+		return nil
+	}
+
+	intervalStr := interval.FormatDuration(period)
+	batch := make(model.Batch, 0, size)
+	for key, aggKey := range current.m {
+		m := makeMetricset(aggKey, intervalStr)
+		batch = append(batch, m)
+		delete(current.m, key)
+	}
+	a.config.Logger.Debugf("publishing %d metricsets", len(batch))
+	return a.config.BatchProcessor.ProcessBatch(ctx, &batch)
+}
+
+// ProcessBatch aggregates all service summary metrics.
+func (a *Aggregator) ProcessBatch(ctx context.Context, b *model.Batch) error {
+	a.mu.RLock()
+	defer a.mu.RUnlock()
+	for _, event := range *b {
+		// Not filtering on Processor.
+		a.processEvent(&event, b)
```

Comment:
Are there any events that we need to exclude? There is absolutely 0 filtering now

Reply:
strictly thinking we could ignore spans, they are only adding noise and no value, but they do not increase cardinality, so also not a big issue to not have a filter at all

Reply:
Done, thanks.